src/ring_theory/ideal_operations.lean

```diff
@@ -321,6 +321,201 @@ or.cases_on (lt_or_eq_of_le $ nat.le_of_lt_succ H)
+    specialize @ih a b (insert i u) this _ _,
+    { refine ih.imp id (or.imp id (exists_imp_exists $ λ k, _)), simp only [exists_prop],
+      exact and.imp (λ hk, finset.insert_subset_insert i (finset.subset_insert j u) hk) id },
+    { rwa finset.card_insert_of_not_mem at hn ⊢,
+      { exact mt (λ hiu, finset.mem_insert_of_mem hiu) hit }, { exact hju } },
+    { change (f j : set R) ≤ f i at hfji,
+      rw finset.coe_insert at h ⊢, rw finset.coe_insert at h,
+      simp only [set.bUnion_insert] at h ⊢,
+      rwa [← set.union_assoc ↑(f i), set.union_eq_self_of_subset_right hfji] at h } },
+  by_cases Ha : f a ≤ f i,
+  { specialize @ih i b t hp.2 hn _,
+    { rcases ih with ih | ih | ⟨k, hkt, ih⟩,
+      { right, right, exact ⟨i, finset.mem_insert_self i t, ih⟩ },
+      { right, left, exact ih },
+      { right, right, exact ⟨k, finset.mem_insert_of_mem hkt, ih⟩ } },
+    { rwa [finset.coe_insert, set.bUnion_insert, set.union_left_comm,
+          ← set.union_assoc, ← set.union_assoc, set.union_eq_self_of_subset_right Ha] at h } },
+  by_cases Hb : f b ≤ f i,
+  { specialize @ih a i t hp.2 hn _,
+    { rcases ih with ih | ih | ⟨k, hkt, ih⟩,
+      { left, exact ih },
+      { right, right, exact ⟨i, finset.mem_insert_self i t, ih⟩ },
+      { right, right, exact ⟨k, finset.mem_insert_of_mem hkt, ih⟩ } },
+    { rwa [finset.coe_insert, set.bUnion_insert, ← set.union_assoc,
+        set.union_assoc ↑(f a), set.union_eq_self_of_subset_left Hb] at h } },
+  by_cases Hi : I ≤ f i,
+  { right, right, exact ⟨i, finset.mem_insert_self i t, Hi⟩ },
+  have : ¬I ⊓ f a ⊓ f b ⊓ t.inf f ≤ f i,
+  { simp only [hp.1.inf_le, hp.1.inf_le', not_or_distrib], exact ⟨⟨⟨Hi, Ha⟩, Hb⟩, Ht⟩ },
+  replace this := set.not_subset.1 this, rcases this with ⟨r, ⟨⟨⟨hrI, hra⟩, hrb⟩, hr⟩, hri⟩,
+  by_cases HI : (I : set R) ⊆ ↑(f a) ∪ ↑(f b) ∪ ⋃ i ∈ (↑t : set ι), f i,
+  { specialize ih hp.2 hn HI, rcases ih with ih | ih | ⟨k, hkt, ih⟩,
+    { left, exact ih }, { right, left, exact ih },
+    { right, right, exact ⟨k, finset.mem_insert_of_mem hkt, ih⟩ } },
+  exfalso, replace HI := set.not_subset.1 HI, rcases HI with ⟨s, hsI, hs⟩,
+  rw [finset.coe_insert, set.bUnion_insert] at h,
+  have hsi : s ∈ f i,
+  { rcases h hsI with ⟨hsa | hsb⟩ | hsi | hst,
+    { exact (hs $ or.inl $ or.inl hsa).elim },
+    { exact (hs $ or.inl $ or.inr hsb).elim },
+    { exact hsi },
+    { exact (hs $ or.inr hst).elim } },
+  rcases h (I.add_mem hrI hsI) with ⟨ha | hb⟩ | hi | ht,
+  { exact hs (or.inl $ or.inl $ add_sub_cancel' r s ▸ (f a).sub_mem ha hra) },
+  { exact hs (or.inl $ or.inr $ add_sub_cancel' r s ▸ (f b).sub_mem hb hrb) },
+  { exact hri (add_sub_cancel r s ▸ (f i).sub_mem hi hsi) },
+  { rw set.mem_bUnion_iff at ht, rcases ht with ⟨j, hjt, hj⟩,
+    simp only [finset.inf_eq_infi, submodule.mem_coe, submodule.mem_infi] at hr,
+    exact hs (or.inr $ set.mem_bUnion hjt $ add_sub_cancel' r s ▸ (f j).sub_mem hj $ hr j hjt) }
+end
+
```

Comment:
@digama0 What do you think about proofs this long? The style is not the style I prefer, I like long proofs to be a series of `have`s. Is this going to be a pain to maintain if the library changes slightly?